### PR TITLE
feat(core-api): hasTransactionFinality() on connector API

### LINKED ISSUE #354: feat(core-api): hasTransactionFinality(): boolean on connector API
**Is your feature request related to a problem? Please describe.**

Need a way to ask the connector at runtime whether its ledger has transaction finality guaranteed or not.

**Describe the solution you'd like**

**Depends on #355**

A method on the connector that returns a boolean, something like:

For ledgers that only support one consensus algorithm (public permissionless ledgers mostly I guess?)

```typescript
public hasTransactionFinality(): Promise<boolean> {
    return false;
}
```

In cases where the ledger has configurable consensus the function's return value would depend on which one is configured for the current ledger instance.

```typescript
public async hasTransactionFinality(): Promise<boolean> {
    const algorithmsWithTxFinality = ["IBFT2", "RAFT", "Other_Algo"];
    const algorithmsWithOutTxFinality = ["Proof_of_Work", "Some_Other_Algo"];
    const currentAlgo = await this.getConsensusAlgorithm();
    const unrecognizedAlgorithm = !algorithmsWithTxFinality.includes(currentAlgo) &&  !algorithmsWithOutTxFinality.includes(currentAlgo);
    if (unrecognizedAlgorithm) { // fail fast when algorithm is unrecognized
        throw new Error(`Unrecognized consensus algorithm: ${currentAlgo}`);
    } else {
        return algorithmsWithTxFinality.includes(currentAlgo);
    }
}
```

To avoid the problem of a ledger introducing a new consensus algorithm that does support TX finality and then us returning incorrect results based on that, we must check the `currentAlgo` for presence in both lists and throw if it isn't present in either one of them. This guarantees that we fail fast instead of non-deterministic behavior.

**Describe alternatives you've considered**

Another way of going about this would be to have the consensus algorithms defined in the consortium definition which would provide us with the means to have the consensus algorithm locked down. Not sure if we should do this because the problem is regarding enforcement: Big mistakes can still happen if a consortium member accidentally changes the consensus algorithm of their own ledger without seeking an update in the consensus definition first.

cc: @sfuji822 @takeutak @jonathan-m-hamilton

### PR DESCRIPTION
**Pull request by jscode017**
Adding function hasTransactionFinality() on connector API, to let ledger query whether the configured consensus has transaction finality at run time.
Fixes #354

### CODE REVIEW

**Comment by petermetz** [comment]
@jscode017 That test failure is shown by the CI logs is not related to the changes you made, if you rebase your branch onto #810 (and mark it as a dependency of this PR) then the tests should start passing. Alternatively just wait a little until #810 gets merged and then rebase onto `main` to achieve the same outcome.

File: packages/cactus-plugin-ledger-connector-besu/src/main/typescript/plugin-ledger-connector-besu.ts
Comment (petermetz) [inline packages/cactus-plugin-ledger-connector-besu/src/main/typescript/plugin-ledger-connector-besu.ts:239]: @jscode017 I would cache the result of 
```typescript
(Object.values(
      ConsensusAlgorithmFamiliesWithTxFinality,
    ) as string[]).includes(currentConsensusAlgorithmFamily.toString())
```

So that at the end it can be returned as-is without further computation needed.

File: packages/cactus-plugin-ledger-connector-corda/src/main/typescript/plugin-ledger-connector-corda.ts
Comment (petermetz) [inline packages/cactus-plugin-ledger-connector-corda/src/main/typescript/plugin-ledger-connector-corda.ts:73]: @jscode017 The part below could be pulled out into a utility function that could be placed in the `cactus-core` package (not `cactus-core-api`).

```typescript
  const unrecognizedConsensusAlgorithmFamily =
      !(Object.values(
        ConsensusAlgorithmFamiliesWithTxFinality,
      ) as string[]).includes(currentConsensusAlgorithmFamily.toString()) &&
      !(Object.values(
        ConsensusAlgorithmFamiliesWithOutTxFinality,
      ) as string[]).includes(currentConsensusAlgorithmFamily.toString());

    if (unrecognizedConsensusAlgorithmFamily) {
      throw new Error(
        `Unrecognized consensus algorithm family: ${currentConsensusAlgorithmFamily}`,
      );
    }
    return (Object.values(
      ConsensusAlgorithmFamiliesWithTxFinality,
    ) as string[]).includes(currentConsensusAlgorithmFamily.toString());
```

Then the `hasTransactionFinality()` methods could all just reuse that method instead of duplicating the same code in every connector.

**Review by petermetz — CHANGES_REQUESTED** [review]

**Review by takeutak — APPROVED** [review]

**Comment by takeutak** [comment]
LGTM

File: packages/cactus-plugin-ledger-connector-corda/src/main/typescript/plugin-ledger-connector-corda.ts
Comment (jscode017) [reply packages/cactus-plugin-ledger-connector-corda/src/main/typescript/plugin-ledger-connector-corda.ts:73]: Yeah, I agree that pulled it out would be a much better way. 
But which class should I put the function to? Both ConsortiumRepository and PluginRegistry in cactus-core do not seems to be a good place?

Comment (jscode017) [reply packages/cactus-plugin-ledger-connector-corda/src/main/typescript/plugin-ledger-connector-corda.ts:73]: BTW, my wrap up function may look like this:
```
public consensusHasTransactionFinality(consensusAlgorithmFamily:ConsensusAlgorithmFamily) boolean{
    const isInConsensusAlgorithmFamiliesWithTxFinality=(Object.values(     ConsensusAlgorithmFamiliesWithTxFinality ) as string[]).includes(consensusAlgorithmFamily.toString());
    
    const isInConsensusAlgorithmFamiliesWithOutTxFinality=(Object.values(  ConsensusAlgorithmFamiliesWithOutTxFinality) as string[]).includes(consensusAlgorithmFamily.toString());
    
    const unrecognizedConsensusAlgorithmFamily =! isInConsensusAlgorithmFamiliesWithTxFinality && !isInConsensusAlgorithmFamiliesWithOutTxFinality;

    if (unrecognizedConsensusAlgorithmFamily) {
      throw new Error(
        `Unrecognized consensus algorithm family: ${consensusAlgorithmFamily}`,
      );
    }
    return isConsensusAlgorithmFamiliesWithTxFinality;
    }
```

**Comment by petermetz** [comment]
@jscode017 Rebasing onto `main` should fix that CI test failure (it is not caused by the changes you made)

File: packages/cactus-plugin-ledger-connector-corda/src/main/typescript/plugin-ledger-connector-corda.ts
Comment (jscode017) [reply packages/cactus-plugin-ledger-connector-corda/src/main/typescript/plugin-ledger-connector-corda.ts:73]: Sorry I get a stupid question here.
I add 
`export { consensusHasTransactionFinality } from "./consensus-has-transaction-finality";`
or
`export * from "./consensus-has-transaction-finality";`
But keep getting
`src/main/typescript/public-api.ts(7,49): error TS2307: Cannot find module './consensus-has-transaction-finality' or its corresponding type declarations.`

Comment (petermetz) [reply packages/cactus-plugin-ledger-connector-corda/src/main/typescript/plugin-ledger-connector-corda.ts:73]: Hmm, that's strange indeed. Can you push to the remote so I can check it out as well?

Comment (jscode017) [reply packages/cactus-plugin-ledger-connector-corda/src/main/typescript/plugin-ledger-connector-corda.ts:73]: Sure , I've pushed it to the remote.

Comment (petermetz) [reply packages/cactus-plugin-ledger-connector-corda/src/main/typescript/plugin-ledger-connector-corda.ts:73]: Ran the build on it, to me it was saying that you are missing `ConsensusAlgorithmFamily` from your import statement (which then makes that file not compile and that leads to the compilation of public-api.ts bugging out.

This is how I got it to build:

```typescript
import {
  ConsensusAlgorithmFamiliesWithTxFinality,
  ConsensusAlgorithmFamiliesWithOutTxFinality,
  ConsensusAlgorithmFamily,
} from "@hyperledger/cactus-core-api";

export function consensusHasTransactionFinality(
  consensusAlgorithmFamily: ConsensusAlgorithmFamily,
): boolean {
  const isInConsensusAlgorithmFamiliesWithTxFinality = (Object.values(
    ConsensusAlgorithmFamiliesWithTxFinality,
  ) as string[]).includes(consensusAlgorithmFamily.toString());

  const isInConsensusAlgorithmFamiliesWithOutTxFinality = (Object.values(
    ConsensusAlgorithmFamiliesWithOutTxFinality,
  ) as string[]).includes(consensusAlgorithmFamily.toString());

  const unrecognizedConsensusAlgorithmFamily =
    !isInConsensusAlgorithmFamiliesWithTxFinality &&
    !isInConsensusAlgorithmFamiliesWithOutTxFinality;

  if (unrecognizedConsensusAlgorithmFamily) {
    throw new Error(
      `Unrecognized consensus algorithm family: ${consensusAlgorithmFamily}`,
    );
  }
  return isInConsensusAlgorithmFamiliesWithTxFinality;
}
//export defualt consensusHasTransactionFinality;

```

Comment (jscode017) [reply packages/cactus-plugin-ledger-connector-corda/src/main/typescript/plugin-ledger-connector-corda.ts:73]: Almost finish.
But there is a small issue remains.
Would want to know how to add dependencies for this project.
I have uploaded a new commit, but somehow cound not refactor "cactus-plugin-ledger-connector-corda".
I would want like to add "@hyperledger/cactus-core-api": "0.4.1" in cactus-plugin-ledger-connector-corda/package.json. 
But instead break dependencies and got bunch of things like below:
```
src/main/typescript/plugin-factory-ledger-connector.ts(4,8): error TS2307: Cannot find module '@hyperledger/cactus-core-api' or its corresponding type declarations.
src/main/typescript/plugin-ledger-connector-corda.ts(15,8): error TS2307: Cannot find module '@hyperledger/cactus-core-api' or its corresponding type declarations.
src/main/typescript/plugin-ledger-connector-corda.ts(16,10): error TS2305: Module '"../../../node_modules/@hyperledger/cactus-core/dist/types/main/typescript"' has no exported member 'consensusHasTransactionFinality'.
src/main/typescript/plugin-ledger-connector-corda.ts(22,8): error TS2307: Cannot find module '@hyperledger/cactus-common' or its corresponding type declarations.
src/main/typescript/plugin-ledger-connector-corda.ts(55,27): error TS2339: Property 'instanceId' does not exist on type 'IPluginLedgerConnectorCordaOptions'.
src/main/typescript/plugin-ledger-connector-corda.ts(60,36): error TS2339: Property 'instanceId' does not exist on type 'IPluginLedgerConnectorCordaOptions'.
src/main/typescript/public-api.ts(15,39): error TS2307: Cannot find module '@hyperledger/cactus-core-api' or its corresponding type declarations.
src/main/typescript/public-api.ts(21,43): error TS2554: Expected 0 arguments, but got 1.
src/main/typescript/web-services/deploy-contract-jars-endpoint.ts(11,8): error TS2307: Cannot find module '@hyperledger/cactus-core-api' or its corresponding type declarations.
src/main/typescript/web-services/deploy-contract-jars-endpoint.ts(18,8): error TS2307: Cannot find module '@hyperledger/cactus-common' or its corresponding type declarations.
src/test/typescript/integration/deploy-cordapp-jars-to-nodes.test.ts(4,33): error TS2307: Cannot find module '@hyperledger/cactus-test-tooling' or its corresponding type declarations.
src/test/typescript/integration/deploy-cordapp-jars-to-nodes.test.ts(5,30): error TS2307: Cannot find module '@hyperledger/cactus-common' or its corresponding type declarations.
src/test/typescript/integration/deploy-cordapp-jars-to-nodes.test.ts(9,8): error TS2307: Cannot find module '@hyperledger/cactus-test-tooling' or its corresponding type declarations.
src/test/typescript/integration/jvm-kotlin-spring-server.test.ts(7,8): error TS2307: Cannot find module '@hyperledger/cactus-test-tooling' or its corresponding type declarations.
src/test/typescript/integration/jvm-kotlin-spring-server.test.ts(8,30): error TS2307: Cannot find module '@hyperledger/cactus-common' or its corresponding type declarations.
src/test/typescript/integration/jvm-kotlin-spring-server.test.ts(12,8): error TS2307: Cannot find module '@hyperledger/cactus-test-tooling' or its corresponding type declarations.
```

File: packages/cactus-core/src/main/typescript/consensus-has-transaction-finality.ts
Comment (petermetz) [inline packages/cactus-core/src/main/typescript/consensus-has-transaction-finality.ts:29]: @jscode017 Recommend to remove commented lines unless you really need them for some specific reason (assuming here this was just forgotten and no longer needed at all)

File: packages/cactus-core/src/main/typescript/public-api.ts
Comment (petermetz) [inline packages/cactus-core/src/main/typescript/public-api.ts:7]: @jscode017 Recommendation here is to explicitly export not with the `*` character (easier to read what exactly goes in the public API that way because then there's no need to open the file up itself and skim through it's exports).
There's exceptions to this practice when there's a lot of exports for example in a generated file, but here it's just the one so I'd say it's easier on the reader if you export it explicitly

```typescript
export { theFunction } from "./consensus-has-transaction-finality";
```

**Review by petermetz — CHANGES_REQUESTED** [review]

File: packages/cactus-core/src/main/typescript/consensus-has-transaction-finality.ts
Comment (jscode017) [reply packages/cactus-core/src/main/typescript/consensus-has-transaction-finality.ts:29]: My bad, forget to do a little clean up

File: packages/cactus-cmd-api-server/src/test/typescript/fixtures/plugin-ledger-connector-stub/plugin-ledger-connector-stub.ts
Comment (petermetz) [reply packages/cactus-cmd-api-server/src/test/typescript/fixtures/plugin-ledger-connector-stub/plugin-ledger-connector-stub.ts:134]: It's perfectly fine as it is IMO. The stub is just there to help with testing so the expectation is that if someone wants to use it to simulate different consensus algorithm families then they can either sub-class it or override the default behavior some other way.

The only thing I need now is the commits to be squashed into a single one so that there are no in-between states (every commit on the main branch should be valid, passing, tests, etc.
So if you squash and then force push (make sure to have the commit signed off) then I'll smash that approve button right after that. Thank you for the contribution and working with me on hammering out the details!

Comment (jscode017) [reply packages/cactus-cmd-api-server/src/test/typescript/fixtures/plugin-ledger-connector-stub/plugin-ledger-connector-stub.ts:134]: Thanks for the help for my first pull request!

Comment (petermetz) [reply packages/cactus-cmd-api-server/src/test/typescript/fixtures/plugin-ledger-connector-stub/plugin-ledger-connector-stub.ts:134]: @jscode017 About to approve and merge this, but I wanted to give one more piece of advice regarding the mechanics of how PR reviews are done:
If someone reviews the PR (e.g. when I literally click button submit review) from that point on it won't notify me until you hit the 're-request review' button (that little refresh icon next to my name in the top right corner among the reviewers).

This is not an actual issue, but I wanted to explain this anyway because I just saw by chance that you had fixed everything that I was asking for more than a week ago but I only noticed it now (otherwise I would've approved+merged the same day)

**Review by petermetz — APPROVED** [review]

File: packages/cactus-cmd-api-server/src/test/typescript/fixtures/plugin-ledger-connector-stub/plugin-ledger-connector-stub.ts
Comment (jscode017) [reply packages/cactus-cmd-api-server/src/test/typescript/fixtures/plugin-ledger-connector-stub/plugin-ledger-connector-stub.ts:134]: Sorry for that, I would hit the 're-request review' next time.

Comment (petermetz) [reply packages/cactus-cmd-api-server/src/test/typescript/fixtures/plugin-ledger-connector-stub/plugin-ledger-connector-stub.ts:134]: No worries at all, thank you again for the PR!

**Pull request merged**
